src/vivarium_public_health/metrics/stratification.py

```diff
@@ -0,0 +1,353 @@
+from dataclasses import dataclass
+from enum import Enum
+import itertools
+from typing import Callable, Dict, Iterable, List, Set, Tuple, Union
+
+import pandas as pd
+from vivarium.framework.engine import Builder
+from vivarium.framework.event import Event
+from vivarium.framework.population import PopulationView
+from vivarium.framework.time import Time
+
+
+class SourceType(Enum):
+    COLUMN = "column"
+    PIPELINE = "pipeline"
+    CLOCK = "clock"
+
+
+@dataclass
+class Source:
+    # todo add docstrings
+    name: str
+    type: SourceType
+
+
+@dataclass
+class StratificationLevel:
+    # todo add docstrings
+    name: str
+    sources: List[Source]
+    categories: Set[str]
+    mapper: Callable[[pd.Series], str] = None
+    current_categories_getter: Callable[[], Set[str]] = None
+
+    def __post_init__(self):
+        self.mapper = self.mapper if self.mapper else self._default_mapper
+        self.current_categories_getter = (
+            self.current_categories_getter if self.current_categories_getter
+            else self._default_current_categories_getter
+        )
+
+    def get_current_categories(self) -> Set[str]:
+        return self.current_categories_getter()
+
+    def _default_mapper(self, row: pd.Series) -> str:
+        category = str(row[0])
+        if category not in self.categories:
+            raise ValueError(f"Invalid value '{category}' found in {self.name}.")
+        return category
+
+    def _default_current_categories_getter(self) -> Set[str]:
+        return self.categories
+
+
+class ResultsStratifier:
+    """Centralized component for handling results stratification.
+
+    This should be used as a subcomponent for observers.  The observers
+    can then ask this component for population subgroups and labels during
+    results production and have this component manage adjustments to the
+    final column labels for the subgroups.
+
+    """
+
+    NAME = "results_stratifier"
```

Comment:
Not sure why you want this

Reply:
Observer components will get the ResultsStratifier by name. This is the constant to be used for that purpose